<!DOCTYPE html>
<html lang="en">

	<head>
		<meta charset="utf-8">
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<title>Riboviz: Methods</title>
		<script type="text/javascript" src="js/jquery-1.11.3.min.js"></script>
		<script type="text/javascript" src="js/jquery.js"></script>
		<script type="text/javascript" src="js/jquery-migrate-1.2.1.min.js"></script>
		<script type="text/javascript" src="js/topojson.v1.min.js"></script>
		<script type="text/javascript" src="js/queue.v1.min.js"></script>
		<script type="text/javascript" src="js/d3.v3.min.js"></script>
		<script type="text/javascript" src="js/bootstrap.min.js"></script>
<!-- 		<script type="text/javascript" src="js/bootstrap.min.css"></script> -->

		<!-- 
<script src="//code.jquery.com/jquery-1.11.3.min.js"></script>
		<script src="http://code.jquery.com/jquery.js"></script>
		<script src="//code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
		<script src="http://d3js.org/topojson.v1.min.js"></script>
		<script src="http://d3js.org/queue.v1.min.js"></script>
		<script type="text/javascript" src="//d3js.org/d3.v3.min.js"></script>
		<script src="http://d3js.org/d3.v3.min.js"></script>
 -->
		<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js" integrity="sha384-0mSbJDEHialfmuBBQP6A4Qrprq5OVfW37PRR3j5ELqxss1yVqOtnepnHVP9aJ7xS" crossorigin="anonymous"></script>
		<link rel="stylesheet"	href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css" integrity="sha384-1q8mTJOASx8j1Au+a5WDVnPi2lkFfwwEAa8hDDdjZlpLegxhjVME1fgjWPGmkzs7" crossorigin="anonymous" media="screen">
			<link href="css/style.css" rel="stylesheet" media="screen">

		<script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.0/jquery.min.js"></script>
		<script src="//netdna.bootstrapcdn.com/bootstrap/3.1.1/js/bootstrap.min.js"></script>
		
		<script type="text/javascript" src="showhide.js"></script>
		
		
			<style type="text/css">
			.axis path,
			.axis line {
  				fill: none;
  				stroke: #000;
  				shape-rendering: crispEdges;
			}

			.axis text {
				font-family: sans-serif;
				font-size: 11px;
			}	
			.x.axis path {
  				display: none;
			}
			.line {
  				fill: none;
  				stroke: rgb(239,138,98);
  				stroke-width: 2px;
			}

			.circle {
  				fill: none;
  				stroke: steelblue;
			}
		</style>	
	</head>
	<body>

		<nav class="navbar navbar-default navbar-static-top" role="navigation">			<div class="container">
				<div class="navbar-header">
					<button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse" aria-expanded="false">
						<span class="sr-only">Toggle navigation</span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
						<span class="icon-bar"></span>
					</button>

					<a class="navbar-brand" href="index.html">
						<div class="navtitle">
							<span class="c4">r</span><span class="c4">i</span><span class="c4">b</span><span class="c4">o</span><span class="c7">v</span><span class="c7">i</span><span class="c7">z</span>
						</div>
					</a>

				</div>
				<div class="collapse navbar-collapse" id="navbar-collapse">
			
					<ul class="nav navbar-nav navbar-right nav-tabs">
						
						<li role="presentation">
						
							<a href="DataPage.html">
								Datasets
							</a>
					
						</li>
						<li role="presentation">
						
							<a href="GeneBased.html">Gene of interest</a>
					
						</li>
						<li role="presentation">
						
							<a href="MethodsPage.html">
								Methods
							</a>
						</li>
						
						<li class="dropdown" role="presentation">
						
							<a href="HelpPage.html" >
								Help	
						</li>

					</ul>
					
				</div>
			

			</div>
		</nav>

	

<div class="container">
<div class="row hidden-xs">
	<div class="col-md-9 paperbox">
		<div class="c10">   Analysis and Visualization of Ribosome  Profiling  Data</div>
	</div>
</div>
</div>


<div class="bigspacer"></div>
<div class="bigspacer"></div>


`<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Methods
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 
		 <p> Analyses of mRNA abundances have been used to gain insight into almost every area of modern biology.  But ultimately, it is protein synthesis that is central to almost every function of the cell. Even though mRNA expression have been used as a proxy for protein production and ultimately molecular function, the correlation between mRNA and protein levels has been shown to be quite weak, likely driven by post-transcriptional regulation. <b>Ribosome profiling (RiboSeq)</b> is a novel technique to study the process and output of translation, the next obvious step following analyses done at the transcription level.  At the core of this technique is the observation that a ribosome translating a fragment of mRNA protects around 30 nucleotides of this mRNA from nuclease activity.  High-throughput sequencing of these ribosome protected fragments (called ribosome footprints) offers a precise record of the number and location of the ribosomes at the time at which translation is stopped. Mapping the position of the ribosome protected fragments indicates the translated regions within the transcriptome.  Moreover, ribosomes spend different periods of time at different positions, leading to variation in the footprint density along the mRNA transcript. This provides an estimate of how much protein is being produced from each mRNA. Importantly, ribosome profiling is as precise and detailed as RNA sequencing. Even in a short time, since its introduction in 2009, ribosome profiling has been playing a key role in driving biological discovery. </p>
		
	<p>	However, the method is not without its limitations.  In mammalian cells, there can be over 10 million unique footprints. The quantification and processing of this set of footprints still remains a challenge and requires computational and domain specific knowledge. One might assume that the similarity of ribosome profiling and RNA-seq allows the use of well established bioinformatic tools with this new type of data. However, ribosome profiling data differs from RNA-seq data in how it is distributed across the genome.  The map and density of ribosomal footprints across the genome contain much more additional information that RNA-seq, which traditional bioinformatics pipelines are not designed to analyze. </p>


<p> We developed a bioinformatic tool-kit, <b> RiboViz</b>, for analyzing and visualizing ribosome profiling data, an essential next step in order for this technology to reach maturation. In implementation,  <b>RiboViz</b> consists of a comprehensive and flexible backend analysis pipeline and a web application for visualization. </p>


<p> Currently, the literature consists of a mix of studies, some which use cycloheximide and some which use flash freezing. In general, each experimental step can potentially cause spurious results ands distortions in the data output. <b> RiboViz</b> can surface potential biases in the data due to experimental abnormalities as well as provide the user the ability to compare and contrast different data sets obtained by different experimental conditions. </p>

<p>  The backend pipeline for processing the raw reads is available GITHUB LINK HERE.
In addition to the metagenomic analyses, a R/Shiny integration allows the user to select a gene of interest and compare ribosomal densities along its ORF with up to three other data sets.
</p>
			
	</div>
</div>
</div>


`<div class="container">		
<div class="row">
	<div class="col-md-3">
		<div class="io-container title">
			Data cleanup and backend processing
		</div>
	</div>
	<div class="col-md-9" align="justify">
		 <p> A major challenge in the analysis of ribosome sequencing data is to be able to map ambiguous reads and remove potential experimental biases that may lead to spurious results using a rigorous quality control processing pipeline. The fact that ribosome footprints have such short sizes preclude any of the longer read or pair end approaches used with other type of sequencing datasets.Profiling datasets typically contain reads from either poly-adenlyated libraries or reads from libraries prepared with a pre-adynlyated linker. These linkers are trimmed using cutadapt and trailing poly-As are removed using custom python scripts based on an iterative procedure described in Ingolia, 2010. Processed reads are mapped to the yeast genome in two steps.  Reads are first mapped to all yeast rRNAs using bowtie with two mismatches allowed per read and all mapped reads will be discarded. All unmapped reads are then be mapped to the yeast genome using bowtie. Reads that map uniquely to a genomic position with at most one mismatch are used for all downstream analyses. All  mapped reads are stored in HDF5 data format using custom R scripts for rapid and easy access to position-specific reads mapping to any gene or sets of genes from an experiment.

At the end of the backend pipeline, HDF5 files are exported containing .... These files can be downloaded... (TO FILL).

<p>
			
	</div>
</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>

<div class="row">
	<div class="col-md-3"></div>
	<div class="col-md-9">
		<hr>
		<div class="center text-gray smallnote">
			&copy; 2016 Oana Carja, Premal Shah
		</div>
	</div>
</div>

<div class="row">
	<div class="col-md-12">
		<div class="bigspacer"></div>
	</div>
</div>



	</body>
</html>

